### PR TITLE
Add uid to events

### LINKED ISSUE #51: ICS export is broken; UID's constantly change. (And add Sequence)
A forum I'm on has just added this plugin, but there is a problem. Every time I hit refresh on https://community.coops.tech/calendar.ics?time_zone=Europe/London the events have a different UID!

This totally breaks things for any system that tries to import the same feed twice, and use the UID to tie data objects together. It's definetly not how an ID is meant to work.

Also, check section 3.8.4.7 of https://tools.ietf.org/html/rfc5545 - note the bit about how many people use @ notation to make unique ID's. The example they give is

           UID:19960401T080045Z-4000F192713-0052@example.com

Can I hazard a guess - you were having problems where an event updated on the site but feed readers did not update it? So you make random UID's to force them to update it? 

The real solution you are looking for is the SEQUENCE property. This should be set on each event, and be updated appropriately. See section 3.8.7.4 of the RFC. This should just be an integer - it doesn't have to go up by one each time, as long as it's an integer which stays the same when there are no changes and goes up by some amount when there are changes, that's great. On one system I did I just took the timestamp of the last updated value for a record and used that as the sequence and it seems fine.

Thanks - the feed at the moment from this plugin can't really be said to be valid ICS, so this is important to fix.

### PR DESCRIPTION
**Pull request by fzngagan**
This fixes #51 

Although, It doesn't add event_version to existing posts. My suggestion is to create a rake task, which goes through all the event posts and adds a `event_version` field if it doesn't exist already.

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -270,43 +276,43 @@ def include_event_going_total?
...
+          tc.topic.custom_fields['event_going_max'] = going_max if going_max_change
 
           going = event['going'] ? event['going'].join(',') : ''
-          tc.record_change('event_going', tc.topic.custom_fields['event_going'], going)
-          tc.topic.custom_fields['event_going'] = going
+          going_change = tc.record_change('event_going', tc.topic.custom_fields['event_going'], going)
+          tc.topic.custom_fields['event_going'] = going if going_change
+        end
+
+        if start_change || end_change || timezone_change # increment by 1, even if more than one props are changed at once
```
Comment (fzngagan) [inline plugin.rb]: Doesn't work if `event_version` field is nil. Serves the purpose of preventing an error

**Comment by fzngagan** [comment]
> @angusmcleod I haven't completed my review, but I've come across an exception which should be fixed first. Steps to reproduce:
> 
> 1. Add an event with a timezone
> 2. Turn on "Agenda" in the category settiings
> 3. Go to http://localhost:3000/c/test/l/agenda.ics
> 4. Get "Invalid identifier" TZInfo::InvalidTimezoneIdentifier exception from discourse-events/plugin.rb line 543

Hi, thanks for reporting this. Turns out this is an issue on the master branch too. I've found the cause of this issue but do we want to add the fix in this PR itself?

File: plugin.rb [hunk 2]
```diff
@@ -563,6 +571,8 @@ def event_ics(opts = {})
             e.summary = t.title
             e.description = t.url << "\n\n" << t.excerpt #add url to event body
             e.url = t.url #most calendar clients don't display this field
+            e.uid = t.id.to_s + "@" + SiteSetting.notification_email.split('@')[1]
```
[the remark this answers is not in the document]
Comment (fzngagan) [reply plugin.rb]: Sure. I'll strip the `protocol` and `www.` from the base_url right?

**Comment by angusmcleod** [comment]
@fzngagan Ah ok, yes let's fix it on this branch. It's necessary to properly test this

**Comment by fzngagan** [comment]
> @fzngagan Ah ok, yes let's fix it on this branch. It's necessary to properly test this

Its actually working. Only a get param `time_zone` needs to be added in the url.
http://localhost:3000/c/events/l/calendar.ics?time_zone=Asia/Calcutta

**Comment by angusmcleod** [comment]
@fzngagan Ah I see ok. Can you create a new open source task on pavilion to use the timezone of the event if no timezone is specified. We don't have to do that right now. I'll review the rest of this shortly.

**Comment by fzngagan** [comment]
> Can you create a new open source task on pavilion to use the timezone of the event if no timezone is specified

Done.

**Pull request merged**